crates/services/p2p/src/peer_manager.rs

```diff
@@ -453,3 +475,217 @@ impl NetworkBehaviour for PeerInfoBehaviour {
+}
+
+impl PeerManager {
+    fn new(reserved_peers: HashSet<PeerId>, max_connections_allowed: usize) -> Self {
+        let reserved_peers_size = reserved_peers.len();
+
+        Self {
+            pending_events: VecDeque::default(),
+            connected_peers: HashMap::with_capacity(max_connections_allowed),
+            reserved_peers,
+            non_reserved_peers_allowed: max_connections_allowed - reserved_peers_size,
+        }
+    }
+
+    fn insert_peer_addresses(&mut self, peer_id: &PeerId, addresses: Vec<Multiaddr>) {
+        if let Some(peer) = self.connected_peers.get_mut(peer_id) {
+            for address in addresses {
+                peer.peer_addresses.insert(address);
+            }
+        } else {
+            log_missing_peer(peer_id);
+        }
+    }
+
+    fn insert_latest_ping(&mut self, peer_id: &PeerId, duration: Duration) {
+        if let Some(peer) = self.connected_peers.get_mut(peer_id) {
+            peer.latest_ping = Some(duration);
+        } else {
+            log_missing_peer(peer_id);
+        }
+    }
+
+    fn insert_client_version(&mut self, peer_id: &PeerId, client_version: String) {
+        if let Some(peer) = self.connected_peers.get_mut(peer_id) {
+            peer.client_version = Some(client_version);
+        } else {
+            log_missing_peer(peer_id);
+        }
+    }
+
+    fn get_disconnected_reserved_peers(&self) -> impl Iterator<Item = &PeerId> {
+        self.reserved_peers
+            .iter()
+            .filter(|peer_id| !self.connected_peers.contains_key(peer_id))
+    }
+
+    fn reserved_peers_connected_count(&self) -> usize {
+        self.reserved_peers.iter().fold(0, |count, peer_id| {
+            if self.connected_peers.contains_key(peer_id) {
+                count + 1
+            } else {
+                count
+            }
+        })
+    }
+
+    fn find_disconnected_reserved_peer(&self) -> Option<PeerId> {
+        self.reserved_peers
+            .iter()
+            .find(|peer_id| self.connected_peers.contains_key(peer_id))
+            .cloned()
+    }
+
+    /// Handles the first connnection established with a Peer
+    fn handle_initial_connection(&mut self, peer_id: PeerId) {
+        // if the connected Peer is not from the reserved peers
+        if !self.reserved_peers.contains(&peer_id) {
+            let non_reserved_peers_connected =
+                self.connected_peers.len() - self.reserved_peers_connected_count();
+
+            // check if there is no more space for non-resereved peers
+            if non_reserved_peers_connected >= self.non_reserved_peers_allowed {
+                // todo/potential improvement: once `Peer Reputation` is implemented we could check if there are peers
+                // with poor reputation and disconnect them instead?
+
+                // Too many peers already connected, disconnect the Peer
+                self.pending_events.push_back(PeerInfoEvent::TooManyPeers {
```

Comment:
Could this be done on handshake rather than after they've (temporarily) consumed a connection slot? If peers are repeatedly reconnecting and consuming available connections, this could explain why there's a big variance in running the reconnect test as it basically has to bruteforce connections until a reserved node gets lucky.

Reply:
Yes, that is very trye, it's just basically waiting for it's turn to connect and stay connected, while other surplus nodes get disconnected. I will experiment a bit with handshake, maybe we could hold some reference to current connection state in that protocol and based on that complete the handshake or reject the connection.

As for the cooldown, I'm not aware there's such a thing, we could have a wrapper over ban/unban if needed